contracts/WooCrossChainRouter.sol

```diff
@@ -1,372 +1,435 @@
+            refId,
+            msgSender,
+            to,
+            srcInfos.fromToken,
+            srcInfos.fromAmount,
+            srcInfos.minBridgeAmount,
+            bridgeAmount
         );
-        uint256 dstGas = (toToken == wooppQuoteTokens[dstChainId]) ? dstGasForNoSwapCall : dstGasForSwapCall;
-        return
-            stargateRouter.quoteLayerZeroFee(
-                dstChainId,
-                1, // https://stargateprotocol.gitbook.io/stargate/developers/function-types
-                toAddress,
-                payloadData,
-                IStargateRouter.lzTxObj(dstGas, 0, "0x")
-            );
     }
 
     function sgReceive(
-        uint16, /*_chainId*/
-        bytes memory, /*_srcAddress*/
-        uint256, /*_nonce*/
-        address _token,
+        uint16, // srcChainId
+        bytes memory, // srcAddress
+        uint256, // nonce
+        address bridgedToken,
         uint256 amountLD,
         bytes memory payload
     ) external override {
-        require(msg.sender == address(stargateRouter), "WooCrossChainRouter: INVALID_CALLER");
+        require(msg.sender == address(stargateRouter), "WooCrossChainRouterV2: INVALID_CALLER");
 
-        (address toToken, uint256 refId, uint256 minToAmount, address to) = abi.decode(
+        // make sure the same order to _encodePayload() when decode payload
+        (uint256 refId, address to, address toToken, uint256 minToAmount) = abi.decode(
             payload,
-            (address, uint256, uint256, address)
+            (uint256, address, address, uint256)
         );
 
-        if (wooRouter.wooPool().quoteToken() != _token) {
-            // NOTE: The bridged token is not WooPP's quote token.
-            // So Cannot do the swap; just return it to users.
-            TransferHelper.safeTransfer(_token, to, amountLD);
+        // when bridged by ETH, the bridgedToken will be SGETH ERC20 address and send native token to address(this)
+        if (toToken == ETH_PLACEHOLDER_ADDR && bridgedToken == sgETHs[sglChainId]) {
+            TransferHelper.safeTransferETH(to, amountLD);
             emit WooCrossSwapOnDstChain(
                 refId,
                 msg.sender,
                 to,
-                _token,
+                ETH_PLACEHOLDER_ADDR,
                 amountLD,
                 toToken,
-                _token,
+                ETH_PLACEHOLDER_ADDR,
                 minToAmount,
                 amountLD
             );
             return;
         }
 
-        uint256 quoteAmount = amountLD;
+        if (toToken == bridgedToken) {
+            TransferHelper.safeTransfer(bridgedToken, to, amountLD);
+            emit WooCrossSwapOnDstChain(
+                refId,
+                msg.sender,
+                to,
+                bridgedToken,
+                amountLD,
+                toToken,
+                bridgedToken,
+                minToAmount,
+                amountLD
+            );
+            return;
+        }
 
```

Comment:
if (bridgedToken == sgETHs[localId]) {
   // sgReceived native
   if (to == PLACEHOLDER) {
      safeTransferETH(...);
   } else {
      wooRouter.swap{value: amountLD};
   }
} else {
   // sgReceived ERC20
   if (....) {
      safeTransfer()
   } else {
     wooRouter.swap
   }
}